File: ios/LuminaLog/Core/Vectors/PeriodIndex.swift
import Foundation

/// Period-index functions for the zoom pyramid's week/month/quarter/year tiers,
/// ported from `server/src/services/periodCentroid/periodIndex.ts`. Must stay
/// byte-for-byte equivalent to that file: both server and client filter entries
/// into the same tiers from the same `dayIndex`, and any drift here silently
/// misfiles a day into the wrong week/month/quarter/year on this client only.
///
/// `dayIndex` (see `ServerSemanticIndex.dayIndex(for:)`) is UTC days-since-epoch.
/// Every function here recovers (y, m, d) with UTC calendar components, exactly
/// like the server's `dateForDayIndex`, so no timezone parameter is needed.
enum PeriodIndex {
    static let lifetimeIndex = 0

    private static var utcCalendar: Calendar = {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        return cal
    }()

    private static func date(forDayIndex dayIndex: Int) -> Date {
        Date(timeIntervalSince1970: TimeInterval(dayIndex) * 86_400)
    }

    /// `year * 12 + zeroBasedMonth`.
    static func monthIndex(forDayIndex dayIndex: Int) -> Int {
        let comps = utcCalendar.dateComponents([.year, .month], from: date(forDayIndex: dayIndex))
        return comps.year! * 12 + (comps.month! - 1)
    }

    /// `year * 4 + zeroBasedQuarter`.
    static func quarterIndex(forDayIndex dayIndex: Int) -> Int {
        let comps = utcCalendar.dateComponents([.year, .month], from: date(forDayIndex: dayIndex))
        return comps.year! * 4 + (comps.month! - 1) / 3
    }

    /// The calendar year.
    static func yearIndex(forDayIndex dayIndex: Int) -> Int {
        utcCalendar.component(.year, from: date(forDayIndex: dayIndex))
    }

    /// The day-index of the Thursday of the ISO week containing `dayIndex`.
    static func thursdayDayIndex(forDayIndex dayIndex: Int) -> Int {
        let d = date(forDayIndex: dayIndex)
        let weekday = utcCalendar.component(.weekday, from: d) // 1 = Sunday ... 7 = Saturday
        let mondayBasedDow = (weekday + 5) % 7 // Mon = 0 .. Sun = 6
        let thursday = utcCalendar.date(byAdding: .day, value: 3 - mondayBasedDow, to: d)!
        return Int(floor(thursday.timeIntervalSince1970 / 86_400))
    }

    /// `isoYear * 100 + isoWeek`, per ISO 8601 (weeks start Monday, week 1 contains
    /// the year's first Thursday).
    static func weekIndex(forDayIndex dayIndex: Int) -> Int {
        let thursdayIdx = thursdayDayIndex(forDayIndex: dayIndex)
        let thursday = date(forDayIndex: thursdayIdx)
        let isoYear = utcCalendar.component(.year, from: thursday)

        var jan4Comps = DateComponents()
        jan4Comps.year = isoYear; jan4Comps.month = 1; jan4Comps.day = 4
        let jan4 = utcCalendar.date(from: jan4Comps)!
        let jan4Weekday = utcCalendar.component(.weekday, from: jan4)
        let jan4MondayBasedDow = (jan4Weekday + 5) % 7
        let week1Monday = utcCalendar.date(byAdding: .day, value: -jan4MondayBasedDow, to: jan4)!

        let days = utcCalendar.dateComponents([.day], from: week1Monday, to: thursday).day!
        let isoWeek = Int((Double(days) / 7.0).rounded()) + 1
        return isoYear * 100 + isoWeek
    }
}

File: web/src/lib/stats/periodIndex.ts
/**
 * Period-index functions for the zoom pyramid's week/month/quarter/year tiers,
 * mirrored from `server/src/services/periodCentroid/periodIndex.ts` and
 * `ios/LuminaLog/Core/Vectors/PeriodIndex.swift`. All three must agree: server,
 * iOS, and web each filter entries into the same tiers from the same `dayIndex`.
 *
 * `dayIndex` is UTC days-since-epoch, matching `dayIndexFor` below and the
 * existing web `localDayKey`/`dayKeyToEpochDays` convention in
 * `web/src/lib/stats/dailyGoalStreak.ts` and `journal/page.tsx`.
 */

/** UTC days-since-epoch for `date`, matching the server's `dateForDayIndex` inverted. */
export function dayIndexFor(date: Date): number {
  return Math.floor(date.getTime() / 86_400_000)
}

function dateForDayIndex(dayIndex: number): Date {
  return new Date(dayIndex * 86_400_000)
}

/** `year * 12 + zeroBasedMonth`, unique and monotonically increasing across years. */
export function monthIndexFromDayIndex(dayIndex: number): number {
  const d = dateForDayIndex(dayIndex)
  return d.getUTCFullYear() * 12 + d.getUTCMonth()
}

/** `year * 4 + zeroBasedQuarter`. */
export function quarterIndexFromDayIndex(dayIndex: number): number {
  const d = dateForDayIndex(dayIndex)
  return d.getUTCFullYear() * 4 + Math.floor(d.getUTCMonth() / 3)
}

/** The calendar year. */
export function yearIndexFromDayIndex(dayIndex: number): number {
  return dateForDayIndex(dayIndex).getUTCFullYear()
}

/**
 * `isoYear * 100 + isoWeek`, per ISO 8601 (weeks start Monday, week 1 contains the
 * year's first Thursday). Encodes the ISO week-numbering year, not the calendar year,
 * so a late-December or early-January date can belong to the adjacent year's week 1
 * or week 52/53.
 */
export function weekIndexFromDayIndex(dayIndex: number): number {
  const d = dateForDayIndex(dayIndex)
  const mondayBasedDow = (d.getUTCDay() + 6) % 7 // Mon=0 .. Sun=6
  const thursday = new Date(d)
  thursday.setUTCDate(d.getUTCDate() - mondayBasedDow + 3)

  const isoYear = thursday.getUTCFullYear()
  const jan4 = new Date(Date.UTC(isoYear, 0, 4))
  const jan4MondayBasedDow = (jan4.getUTCDay() + 6) % 7
  const week1Monday = new Date(jan4)
  week1Monday.setUTCDate(jan4.getUTCDate() - jan4MondayBasedDow)

  const isoWeek = Math.round((thursday.getTime() - week1Monday.getTime()) / (7 * 86_400_000)) + 1
  return isoYear * 100 + isoWeek
}

/**
 * The day-index of the Thursday of the ISO week containing `dayIndex`. Used to derive
 * a week's calendar month/quarter/year attribution from a fixed point that cannot
 * flip depending on which day in the week triggered the computation.
 */
export function thursdayDayIndexFromDayIndex(dayIndex: number): number {
  const d = dateForDayIndex(dayIndex)
  const mondayBasedDow = (d.getUTCDay() + 6) % 7
  const thursday = new Date(d)
  thursday.setUTCDate(d.getUTCDate() - mondayBasedDow + 3)
  return Math.floor(thursday.getTime() / 86_400_000)
}
